onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorGroupQueryAttention.cpp

```diff
@@ -0,0 +1,315 @@
+                intermediateEdges.push_back(castToMhaEdge);
+            }
+        }
+        else
+        {
+            // Link the query/key/value inputs to MHA
+            for (uint32_t i = 0; i < 3; ++i)
+            {
+                DML_INPUT_GRAPH_EDGE_DESC inputToMhaEdge = {};
+                inputToMhaEdge.GraphInputIndex = i;
+                inputToMhaEdge.ToNodeIndex = 0;
+                inputToMhaEdge.ToNodeInputIndex = i;
+                inputEdges.push_back(inputToMhaEdge);
+            }
+        }
+
+        constexpr uint32_t dmlPastSequenceLengthsIndex = 11;
+
+        // The GQA offline fusion does this thing where it sums the number of 1's in the mask to figure out the value of the past sequence.
+        // This doesn't work well for the first iteration since, obviously, there are no past sequences and the mask in this case represents
+        // only the elements in the initial sequence. To work around this, the CUDA implementation of the operator ignores the value of
+        // pastSequenceLengths for the first iteration and acts as if it was 0. This feels like a pretty dirty hack and something that should
+        // be polished in the future, but for compatibility with the GQA fusion and the CUDA implementation we do the same thing here. We DO NOT
+        // want to do this within DirectML since DirectML should be agnostic w.r.t which iteration it's currently executing MHA for, and such a
+        // hack that is likely to be modified in the future shouldn't be enshrined within DirectML. Doing it here is OK because the nature of contrib
+        // ops is that they can change at any time.
+        if (sequenceLength == 1)
+        {
+            // Link the PastSequenceLengths input to MHA
+            DML_INPUT_GRAPH_EDGE_DESC inputToMhaEdge = {};
+            inputToMhaEdge.GraphInputIndex = seqLensIndex;
+            inputToMhaEdge.ToNodeIndex = 0;
+            inputToMhaEdge.ToNodeInputIndex = dmlPastSequenceLengthsIndex;
+            inputEdges.push_back(inputToMhaEdge);
+        }
+        else
+        {
+            opDescs.push_back(&zeroScalarDmlDesc);
+
+            // Link the zero scalar to MHA
+            DML_INTERMEDIATE_GRAPH_EDGE_DESC zeroScalarToMhaEdge = {};
+            zeroScalarToMhaEdge.FromNodeIndex = gsl::narrow_cast<uint32_t>(opDescs.size() - 1);
+            zeroScalarToMhaEdge.FromNodeOutputIndex = 0;
+            zeroScalarToMhaEdge.ToNodeIndex = 0;
+            zeroScalarToMhaEdge.ToNodeInputIndex = dmlPastSequenceLengthsIndex;
+            intermediateEdges.push_back(zeroScalarToMhaEdge);
+        }
+
+        if (isFp16)
+        {
+            // Link MHA's output to the output cast nodes
+            for (uint32_t i = 0; i < 3; ++i)
+            {
+                DML_INTERMEDIATE_GRAPH_EDGE_DESC mhaToCastEdge = {};
+                mhaToCastEdge.FromNodeIndex = 0;
+                mhaToCastEdge.FromNodeOutputIndex = i;
+                mhaToCastEdge.ToNodeIndex = 4 + i;
```

Comment:
Hmm, can you explain where the `4+` comes from? (the counted loop has 3 items in it, and it's not immediately apparent to me and probably future readers the significance of `4`).